### ISSUE TITLE
Add a hook applications can use to determine the connection destination given the request URL

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#6634]
|[<img alt="habnabit's avatar" src="https://avatars.githubusercontent.com/u/21168?s=50" width="50" height="50">](https://github.com/habnabit)| @habnabit reported|
|-|-|
|Trac ID|trac#6634|
|Type|enhancement|
|Created|2013-07-19 16:24:54Z|
|Branch|https://github.com/twisted/twisted/tree/agent-endpointFactory-6634-2|

Currently, there's a `_getEndpoint` method, but it's a private implementation detail. This could be turned into a parameter to `Agent.__init__` so that inheriting from Agent isn't necessary to specify an endpoint factory.

<details><summary>Searchable metadata</summary>

```
trac-id__6634 6634
type__enhancement enhancement
reporter__habnabit habnabit
priority__normal normal
milestone__ 
branch__branches_agent_endpointFactory_6634_2 branches/agent-endpointFactory-6634-2
branch_author__habnabit habnabit
status__closed closed
resolution__fixed fixed
component__web web
keywords__ 
time__1374251094000000 1374251094000000
changetime__1400619546000000 1400619546000000
version__None None
owner__itamar itamar
cc__jknight
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_2"></a>|@exarkun commented|
|-|-|

"Callable of X, Y, Z" is an awkward, inconvenient, and non-extensible interface.  Please avoid introducing new APIs that work this way.

Also please describe why this might be useful.  After all, Agent already knows how to create an endpoint for both HTTP and HTTPS, both of the kinds of things.

**Comment by twisted-trac** [comment]
|[<img alt="habnabit's avatar" src="https://avatars.githubusercontent.com/u/21168?s=50" width="50" height="50">](https://github.com/habnabit)<a name="note_3"></a>|@habnabit commented|
|-|-|

(In [39263]) Branching to 'agent-endpointFactory-6634'

**Comment by twisted-trac** [comment]
|[<img alt="habnabit's avatar" src="https://avatars.githubusercontent.com/u/21168?s=50" width="50" height="50">](https://github.com/habnabit)<a name="note_4"></a>|@habnabit commented|
|-|-|

(In [39264]) Add endpointFactory to twisted.web.client.Agent.

Refs [#6634](https://github.com/twisted/twisted/issues/6634).

**Comment by twisted-trac** [comment]
|[<img alt="habnabit's avatar" src="https://avatars.githubusercontent.com/u/21168?s=50" width="50" height="50">](https://github.com/habnabit)<a name="note_5"></a>|@habnabit commented|
|-|-|

(In [42600]) Branching to 'agent-endpointFactory-6634-2'

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_6"></a>|@exarkun commented|
|-|-|

> Also please describe why this might be useful. After all, Agent already knows how to create an endpoint for both HTTP and HTTPS, both of the kinds of things.

Additionally, `ProxyAgent.__init__` accepts an endpoint.

**Comment by twisted-trac** [comment]
|[<img alt="habnabit's avatar" src="https://avatars.githubusercontent.com/u/21168?s=50" width="50" height="50">](https://github.com/habnabit)<a name="note_7"></a>|@habnabit commented|
|-|-|

Replying to [exarkun](#note_6):
> > Also please describe why this might be useful. After all, Agent already knows how to create an endpoint for both HTTP and HTTPS, both of the kinds of things.
> 
> Additionally, `ProxyAgent.__init__` accepts an endpoint.
`ProxyAgent` only makes HTTP proxy requests. In the case of, say, connecting to a server over SOCKS5, ultimately you want to issue a regular request to the remote end, and not an HTTP proxy request.

Right now there's two distinct use cases that I can think of for this:
 1. Connecting to a server over a proxy mechanism like SOCKS5. Right now there is no way to do this at all with the public API for `Agent`.
 2. Connecting to a server over a different transport type, like UNIX domain sockets. Docker is one example of software that does HTTP requests over UNIX domain sockets, and it would be useful to be able to communicate with it without having to bind to a TCP port. Again, right now there is no way to do this at all with the public API for `Agent`.

**Comment by twisted-trac** [comment]
|[<img alt="habnabit's avatar" src="https://avatars.githubusercontent.com/u/21168?s=50" width="50" height="50">](https://github.com/habnabit)<a name="note_8"></a>|@habnabit commented|
|-|-|

(In [42604]) Reconstruct everything from scratch.

The old branch diverged way too much and wasn't very good in the first place.
So, start over.

refs [#6634](https://github.com/twisted/twisted/issues/6634)

**Comment by twisted-trac** [comment]
|[<img alt="habnabit's avatar" src="https://avatars.githubusercontent.com/u/21168?s=50" width="50" height="50">](https://github.com/habnabit)<a name="note_9"></a>|@habnabit commented|
|-|-|

[Forced a build](http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/agent-endpointFactory-6634-2).

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_10"></a>|@itamarst set owner to @itamarst|
|-|-|

I see two problems with this design:

Having some arguments in `Agent` ignored if other options are passed in is a bit of a hack. I therefore propose we add a new class `EndpointAgent` which takes (in addition to reactor and pool) an `endpointConstructor` argument. 

This could either be something that provides the currently proposed interface, or a simpler variant. I feel like the `IAgentEndpointConstructor` interface is heavily optimized for one use case (SOCKS) but not the other (Docker unix socket).

Specifically, a simpler version would have the endpoint constructor be a two argument callable that takes a URL and reactor and returns an endpoint. The Unix socket constructor would just be `lambda url: UnixEndpoint(mypath, reactor)`. The SOCKS one would need to be constructed with a relevant TLS argument and do URL parsing ... but I bet we could provide a utility that does that if there's more than one endpoint constructor that needs to implement that particular code.

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_11"></a>|@itamarst set owner to @habnabit|
|-|-|

After noting exarkun's comment above:

1. Move same functionality into new class, `EndpointAgent`.

2. Optionally, simpler example of using a Unix socket (to talk to Docker? that'd be a realistic "you can run it at home" example).

3. Classes implementing an interface should use `@implementer(IFoo)`... and don't need to copy the whole docstring since it's in the interface.

Resubmit for review and I'll try to do a final pass.

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_13"></a>|@itamarst set owner to @exarkun|
|-|-|

Started new buildbot run (but it's still running so may be red at end). Otherwise ready for review.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun set status to `assigned`|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_15"></a>|@exarkun set owner to @itamarst<br>@exarkun set status to `new`|
|-|-|

Thanks.

  1. docs/web/howto/listings/client/endpointconstructor.py
    1. Needs a copyright header
    1. This particular example seems like it would be better satisfied by `ProxyAgent`.  Or do I _still_ not understand what `ProxyAgent` does/is for (wish it had some docs)?  Maybe a good thing to include in this branch would be some docs differentiating these two things for dense people like me.
    1. Richard built us some tools for testing examples.  I don't think we decided yet that this means we can finally include examples in the "everything must have tests" policy - but it seems like maybe we should?  Not a blocker but if you don't do that here at least file a ticket for doing it.
  1. docs/web/howto/client.rst
    1. APIs in docs should be `:api:...` links - eg `Agent` and     1. An arbitrary endpoint is how you implement some useful functionality.  The useful functionality is controlling what an agent connects to though.  I suggest tweaking the heading and language a bit to be more inviting to people who know what they want to do but not yet how they want to do it?
`Agent.forEndpointConstructor` in the new text.
  1. The news fragment could probably also be tweaked in a similar way.
  1. twisted/web/iweb.py
    1. "Constructor" seems like an unnecessary variation from "Factory".  I know everyone hates "Factory" but I don't think the difference between "Constructor" and "Factory" conveys any interesting nuance of meaning here, the new word is just a potential source of confusion.
    1. Similar comment for "constructEndpoint" - we used "build" elsewhere already (`buildProtocol`) so maybe `buildEndpoint` is better - or some entirely different style of name, like `endpointFor...` oops not sure what noun to put at the end there because what is a "scheme, hostname, port" triple?  Not a URI...  This is more nit-picky and people will probably figure out the current name, so this is optional.
    1. I guess the ideal signature here, though, would be to accept a request object. :(  I'm not sure what interesting stuff you could do with the whole request here but I suspect there's something - client-side session binding?  Interesting routing choices based on the path of the URI?  "/foo/bar" always goes via my tor proxy, "/foo/baz" just gets a regular https connection... sort of thing.  But we have no request object, sad.  Maybe `endpointForURI` and pass the unparsed URI (or the `urlparse.urlparse` parsed form, crummy as it is).  Later we can introduce more `endpointForSomethingElse` methods that get more information and take precedent if they are defined.  Or, I dunno, maybe a `Request` structure and pass it all in.  The current signature, `scheme, hostname, port` seems like it should definitely change somehow, at least.
  1. twisted/web/client.py
    1. `_StandardEndpointConstructor.constructEndpoint` should have a bit of a docstring, something like "Get an endpoint which connects directly to the host given by the URI."  Somewhat implied by "Standard" but being explicit here doesn't hurt (I think!  does it?  if adding any kind of docstring at all introduces pydoctor or twistedchecker errors then maybe it isn't worth it)
    1. Maybe this is a good opportunity to introduce explicit byte string literal syntax in `constructEndpoint`.  Optional.
    1. `Agent.__init__` still accepts (and now ignores) an `endpointConstructor` parameter.  Looks like this should just be deleted.
    1. Regarding the implementation of `Agent.forEndpointConstructor` - bleugh.  I wish there were a better way to do this.  This comment is not actionable.
    1. Nit... does "for" make the most sense here?  What about `with` or `using`?  Optional.
    1. For what it's worth, I really like the amount of code remaining in `Agent` after this change.
  1. twisted/web/test/test_agent.py
    1. `FakeEndpointConstructor` looks like a cross between a spy and a stub.  It also has a bunch of public attributes that aren't part of the interface it implements.
    1. `test_endpointConstructor` uses native string literals in a couple places where it should use byte string literals.
    1. The docstring for `test_endpointConstructor` is a little incomplete.  The test seems primarily to be about the constructor being used to create endpoints.
    1. `test_endpointConstructorDefaultPool` would be easier if `HTTPConnectionPool` implemented comparison for equality.
    1. `test_endpointConstructorPool` uses `assertIdentical`, should use `assertIs`.
    1. It seems that some (existing) tests (that are now) for `_StandardEndpointConstructor` could be factored into a new test case and could test that class directly.  Optional, file a ticket and link to it from someplace in the code if you don't want to work on this now.
  1. Buildbot is mostly green.  There are some twistedchecker errors but they might be bogus.
  1. I notice your employer isn't in the LICENSE file yet.  If you think it is warranted please add a line there.

I'd like to have another look once the issues with the example are addressed (one way or another).  Thanks again.

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_16"></a>|@itamarst commented|
|-|-|

ProxyAgent changes the way requests are done, it does "GET http://yourhost.com/foo.html HTTP/1.1" instead of "GET /foo.html HTTP/1.1" if you go to URL "http://yourhost.com/foo.html".

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_17"></a>|@itamarst commented|
|-|-|

Perhaps we should make `_URI` object public, and then it could be passed to the endpoint factory...

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_18"></a>|@itamarst set owner to @exarkun|
|-|-|

Implemented all required comments, and some of the optional ones. Filed a couple of follow up tickets as well ([#7385](https://github.com/twisted/twisted/issues/7385), [#7386](https://github.com/twisted/twisted/issues/7386)).

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_19"></a>|@exarkun set owner to @itamarst|
|-|-|

  1. docs/web/howto/client.rst: ''that takes a ``endpointFactory``'' - a -> an
  1. docs/web/howto/listings/client/endpointconstructor.py
    1. The UNIX socket path should use byte literal syntax
    1. A while ago I remember there was some effort to make sure examples all had usage documentation at the top.  This isn't an example I guess (it's not even in the examples directory!) but maybe a quick blurb about the argument this expects would be a good thing.  See, for example, `docs/web/examples/getpage.py` for the kind of thing I mean.

That's it.  This looks pretty nice now, thanks for your work itamar and habnabit.  Please address those minor points, get a greenish buildbot run, and then merge.

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_20"></a>|@itamarst set status to `closed`|
|-|-|

(In [42679]) Merge agent-endpointFactory-6634-2: Allow creating Agent that connect with arbitrary endpoints.

Author: habnabit, itamar
Review: itamar, exarkun
Fixes: [#6634](https://github.com/twisted/twisted/issues/6634)

Agent.usingEndpointFactory allows making HTTP connections with arbitrary endpoints, e.g. with a SOCKS proxy or a Unix socket.
